src/include/denc.h

```diff
@@ -1818,8 +1818,13 @@ inline std::enable_if_t<traits::supported && !traits::featured> decode_nohead(
 			  __u8 *struct_compat,				\
 			  char **start_pos,				\
 			  uint32_t *struct_len) {			\
+    __u8 code_v = *struct_v;						\
     denc(*struct_v, p);							\
     denc(*struct_compat, p);						\
+    if (code_v < *struct_compat)					\
+      throw ::ceph::buffer::malformed_input(fmt::format(		\
```

Comment:
IIUC `struct_compat` is a way to let an encoder decide whether a decoder at particular version should refuse-to-decode new bytestream. This is quite fragile as it establishes an interface spanning over the time dimension.

This change would fix only new decoders; in this aspect we don't have any control over the past ones WHILE the interface for current-and-newer encoders still claims we actually do. This can be confusing and dangerous.

In my understanding `DENC` simply doesn't support the `struct_compat` versioning. Shouldn't we be more explicit about that? How about removing the parameter (only **at the interface layer**, bytestreams must be obviously compatible)? Later we could introduce a new macro called e.g. `DENC_WITH_MIN_COMPAT_SINCE_SQUID()`.

Reply:
The interface for DENC_START always claimed to address forward compatibility problem, but it did not do it.
This already is confusing and dangerous.
Introduction of any new macro that produces backward-compatible bitstream works to visually signal developer, but does not protect against old decoders.
If we are to signal developers, existing `DENC_START` should get renamed to `DENC_START_UNSAFE`.

To make version change reliable we need new encoding that poisons old decoders.
It will be used when bumping up version.
That is, if we transition from `v=1` to `v=2`, the new bitstream will kill old decoder, while the new one should be able to read both.

Reply:
>  This already is confusing and dangerous.

Currently it's _consistently_ (across all versions) confusing and dangerous.
After the patch it would become more _surprising_ – the check would _usually work_ (especially on developer's desk) but not _always_.
Maybe altering the interface to honor what's possible to implement and what's not is the way to go.

To what extent honor? An encoder could detect (thankfully to feature bits) whether the decoder handles the compat version. The question is whether  it's really useful. I could imagine an interface (only for `min_compat`-taking variant?) that would both visually inform and enforce to handle (or ignore) broken decoder explicitly. However, this might be too much of complexity.